## Importing Libraries

In [1]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

## Importing Dataframes

In [2]:
# Define path
path = r'/Users/isaacotubanjo/Documents/08:08:2023 Instacart Business Analysis'

In [3]:
# Import orders_products_merged.pkl as df_ords_prods_merged
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

## Removing the first 1,000,000 rows

In [4]:
df = df_ords_prods_merged[:1000000]

In [6]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [7]:
df.shape

(1000000, 14)

## Defining a function and applying it to the df subset

In [9]:
# Define a function
def price_label (row):
    
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] >5) and (row['prices'] <=15):
        return 'Mid range product'
    elif (row['prices'] >15):
        return 'High range product'
    else: return 'Not enough data'

In [10]:
# Applying the function
df['price_range'] = df.apply(price_label, axis=1)
# This will create a new column titled price range

/var/folders/kg/f38q3svs3lgc6yvw293mvtlr0000gn/T/ipykernel_48906/3340479041.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [16]:
# Checking the sum of the various categories created
df['price_range'].value_counts(dropna = False)

Mid range product    756450
Low range product    243550
Name: price_range, dtype: int64

## Using the LOC method

In [17]:
# Applying the LOC function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/kg/f38q3svs3lgc6yvw293mvtlr0000gn/T/ipykernel_48906/3475891532.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [18]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [19]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
# Checking the sum of the various categories created
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

## Applying the LOC method to the entire dataframe

In [24]:
# Applying the LOC function
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [23]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [22]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [25]:
# Checking the sum of the various categories created
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

## Discovering the busiest day(s) of the week using a For-Loop statement

In [28]:
# Discovering the busiest day(s)
result = []

for value in df_ords_prods_merged["order_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [29]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [31]:
# Merging it with the dataframe df_ords_prods_merged
df_ords_prods_merged['busiest_day_of_week'] = result

In [32]:
df_ords_prods_merged['busiest_day_of_week'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day_of_week, dtype: int64

In [33]:
df_ords_prods_merged

order_id  user_id  order_number  order_day_of_week  \
0          2539329        1             1                  2   
1          2398795        1             2                  3   
2           473747        1             3                  3   
3          2254736        1             4                  4   
4           431534        1             5                  4   
...            ...      ...           ...                ...   
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_last_order  product_id  \
0                         8                    NaN         196   
1                         7                   15.0         196   
2                        12                   21.0         196   
3                         7                   29.0         196   
4                        15                   28.0         196   
...                     ...                    ...         ...   
32404854                 15                    1.0       43553   
32404855                 11                    3.0       43553   
32404856                  7                    NaN       42338   
32404857                  5                   15.0       42338   
32404858                 14                    NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge    price_range_loc  \
0               77              7     9.0   both  Mid-range product   
1               77              7     9.0   both  Mid-range product   
2               77              7     9.0   both  Mid-range product   
3               77              7     9.0   both  Mid-range product   
4               77              7     9.0   both  Mid-range product   
...            ...            ...     ...    ...                ...   
32404854        64              7     3.7   both  Low-range product   
32404855        64              7     3.7   both  Low-range product   
32404856        50             19     6.9   both  Mid-range product   
32404857        50             19     6.9   both  Mid-range product   
32404858       130             14    10.3   both  Mid-range product   

         busiest_day_of_week  
0             Regularly busy  
1             Regularly busy  
2             Regularly busy  
3                 Least busy  
4                 Least busy  
...                      ...  
32404854      Regularly busy  
32404855      Regularly busy  
32404856      Regularly busy  
32404857      Regularly busy  
32404858          Least busy  

[32404859 rows x 16 columns]

# Task 4.7

In [34]:
# Removing the merge column from the dataframe as it is not needed
df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['_merge'])

In [35]:
df_ords_prods_merged

order_id  user_id  order_number  order_day_of_week  \
0          2539329        1             1                  2   
1          2398795        1             2                  3   
2           473747        1             3                  3   
3          2254736        1             4                  4   
4           431534        1             5                  4   
...            ...      ...           ...                ...   
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_last_order  product_id  \
0                         8                    NaN         196   
1                         7                   15.0         196   
2                        12                   21.0         196   
3                         7                   29.0         196   
4                        15                   28.0         196   
...                     ...                    ...         ...   
32404854                 15                    1.0       43553   
32404855                 11                    3.0       43553   
32404856                  7                    NaN       42338   
32404857                  5                   15.0       42338   
32404858                 14                    NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices    price_range_loc  \
0               77              7     9.0  Mid-range product   
1               77              7     9.0  Mid-range product   
2               77              7     9.0  Mid-range product   
3               77              7     9.0  Mid-range product   
4               77              7     9.0  Mid-range product   
...            ...            ...     ...                ...   
32404854        64              7     3.7  Low-range product   
32404855        64              7     3.7  Low-range product   
32404856        50             19     6.9  Mid-range product   
32404857        50             19     6.9  Mid-range product   
32404858       130             14    10.3  Mid-range product   

         busiest_day_of_week  
0             Regularly busy  
1             Regularly busy  
2             Regularly busy  
3                 Least busy  
4                 Least busy  
...                      ...  
32404854      Regularly busy  
32404855      Regularly busy  
32404856      Regularly busy  
32404857      Regularly busy  
32404858          Least busy  

[32404859 rows x 15 columns]

## Creating a new column for the two busiest and two slowest days

In [37]:
# Finding out the busiest and slowest days of the week
df_ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

With this, we can establish that Saturday (0) and Sunday (1) are the busiest days of the week and Tuesday (3) and Wednesday (4) are the slowest days of the week.

Saturday and Sunday will be designated as Busiest days 
Wednesday and Thursday will be designated as Least busy days 
All other week days will be designated as Regularly busy day


In [45]:
# Using the for-loop statement
end_result = []

for value in df_ords_prods_merged["order_day_of_week"]:
  if value == 0 or value == 1:
    end_result.append("Busiest days")
  elif value == 4 or value == 3:
    end_result.append("Least busy days")
  else:
    end_result.append("Regularly busy day")

In [46]:
end_result

['Regularly busy day',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Regularly busy day',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Busiest days',
 'Regularly busy day',
 'Regularly busy day',
 'Busiest days',
 'Busiest days',
 'Regularly busy day',
 'Regularly busy day',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy day',
 'Regularly busy day',
 'Busiest days',
 'Regularly busy day',
 'Regularly busy day',
 'Busiest days',
 'Regularly busy day',
 'Regularly busy day',
 'Regularly busy day',
 'Least busy days',
 'Regularly busy day',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy day',
 'Least busy days',
 'Least busy days',
 'Regularly busy day',
 'Regularly busy day',
 'Busiest days',
 'Busiest days',
 'Regularly busy day',
 'Busiest days',
 'Busiest days

In [47]:
# Merging it with the dataframe df_ords_prods_merged
df_ords_prods_merged['busiest_days'] = end_result

In [52]:
# Checking column for accuracy
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy day    12916111
Busiest days          11864412
Least busy days        7624336
Name: busiest_days, dtype: int64

When compared with the sum of the order_day_of_the_week column, the value has remained consistent (32,404,859)

In [50]:
df_ords_prods_merged

order_id  user_id  order_number  order_day_of_week  \
0          2539329        1             1                  2   
1          2398795        1             2                  3   
2           473747        1             3                  3   
3          2254736        1             4                  4   
4           431534        1             5                  4   
...            ...      ...           ...                ...   
32404854   1320836   202557            17                  2   
32404855     31526   202557            18                  5   
32404856    758936   203436             1                  2   
32404857   2745165   203436             2                  3   
32404858   3093936   205420             1                  4   

          order_hour_of_day  days_since_last_order  product_id  \
0                         8                    NaN         196   
1                         7                   15.0         196   
2                        12                   21.0         196   
3                         7                   29.0         196   
4                        15                   28.0         196   
...                     ...                    ...         ...   
32404854                 15                    1.0       43553   
32404855                 11                    3.0       43553   
32404856                  7                    NaN       42338   
32404857                  5                   15.0       42338   
32404858                 14                    NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices    price_range_loc  \
0               77              7     9.0  Mid-range product   
1               77              7     9.0  Mid-range product   
2               77              7     9.0  Mid-range product   
3               77              7     9.0  Mid-range product   
4               77              7     9.0  Mid-range product   
...            ...            ...     ...                ...   
32404854        64              7     3.7  Low-range product   
32404855        64              7     3.7  Low-range product   
32404856        50             19     6.9  Mid-range product   
32404857        50             19     6.9  Mid-range product   
32404858       130             14    10.3  Mid-range product   

         busiest_day_of_week        busiest_days  
0             Regularly busy  Regularly busy day  
1             Regularly busy     Least busy days  
2             Regularly busy     Least busy days  
3                 Least busy     Least busy days  
4                 Least busy     Least busy days  
...                      ...                 ...  
32404854      Regularly busy  Regularly busy day  
32404855      Regularly busy  Regularly busy day  
32404856      Regularly busy  Regularly busy day  
32404857      Regularly busy     Least busy days  
32404858          Least busy     Least busy days  

[32404859 rows x 16 columns]

## Deriving a column for Busiest Period of the day

In [51]:
# Finding out the busiest and slowest hours of the day
df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

Since there are 24 (0-23) variables for hours of the day, I will divide them into 3 sections with 8 values each:
Most orders - 9, 10, 11, 12, 13, 14, 15, 16
Average orders - 7, 8, 17, 18, 19, 20, 21, 22
Fewest orders - 0, 1, 2, 3, 4, 5, 6, 23

In [55]:
# Using the for-loop statement
hour_of_day = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value in [9, 10, 11, 12, 13, 14, 15, 16]:
    hour_of_day.append("Most orders")
  elif value in [0, 1, 2, 3, 4, 5, 6, 23]:
    hour_of_day.append("Fewest orders")
  else:
    hour_of_day.append("Average orders")

In [56]:
hour_of_day

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [57]:
# Merging it with the dataframe df_ords_prods_merged
df_ords_prods_merged['busiest_period_of_day'] = hour_of_day

In [59]:
# Checking the column's frequency and accuracy
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

When compared with the sum of the order_hour_of_day column, the value has remained consistent (32,404,859)

In [60]:
df_ords_prods_merged.shape

(32404859, 17)

## Exporting as a pickle file

In [61]:
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_with_new_variables.pkl'))